src/python/pants/backend/python/subsystems/python_native_code.py

```diff
@@ -129,3 +139,157 @@ def check_build_for_current_platform_only(self, targets):
       'native code. Please ensure that the platform arguments in all relevant targets and build '
       'options are compatible with the current platform. Found targets for platforms: {}'
       .format(str(platforms_with_sources)))
+
+
+@rule(CToolchain, [Select(PythonNativeCode)])
```

Comment:
100% correct on why this is required. This concern was vaguely addressed with the `TODO(#4020): These classes are performing the work of variants.` I can add your comment to the usage here to make it clear how variants would improve the ux.

Reply:
A comment would be great, but this is not just UX, it's a bug. There is currently no way to select GCC.

Reply:
Sure, but in the test you select GCC explicitly. What's the point of the generic CToolchain if you know you're getting an LLVMCToolchain or GCCCToolchain toolchain because you ask for one of those directly by type. There is perhaps more abstraction than needed.